rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+			return enum_node_type::OPIN;
+		break;
+		case onechar('S', 0, 32) | onechar('I', 8, 32) | onechar('N', 16, 32) | onechar('K', 24, 32):
+			return enum_node_type::SINK;
+		break;
+		default: break;
+		}
+		break;
+	case 5:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('C', 0, 32) | onechar('H', 8, 32) | onechar('A', 16, 32) | onechar('N', 24, 32):
+			switch(in[4]){
+			case onechar('X', 0, 8):
+				return enum_node_type::CHANX;
+			break;
+			case onechar('Y', 0, 8):
+				return enum_node_type::CHANY;
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	case 6:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('S', 0, 32) | onechar('O', 8, 32) | onechar('U', 16, 32) | onechar('R', 24, 32):
+			switch(in[4]){
+			case onechar('C', 0, 8):
+				switch(in[5]){
+				case onechar('E', 0, 8):
+					return enum_node_type::SOURCE;
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	default: break;
+	}
+	if(throw_on_invalid)
+		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_node_type.");
+	return enum_node_type::UXSD_INVALID;
+}
+
+inline enum_node_direction lex_node_direction(const char *in, bool throw_on_invalid){
+	unsigned int len = strlen(in);
+	switch(len){
+	case 6:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('B', 0, 32) | onechar('I', 8, 32) | onechar('_', 16, 32) | onechar('D', 24, 32):
+			switch(in[4]){
+			case onechar('I', 0, 8):
+				switch(in[5]){
+				case onechar('R', 0, 8):
+					return enum_node_direction::BI_DIR;
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	case 7:
+		switch(*((triehash_uu32*)&in[0])){
+		case onechar('D', 0, 32) | onechar('E', 8, 32) | onechar('C', 16, 32) | onechar('_', 24, 32):
+			switch(in[4]){
+			case onechar('D', 0, 8):
+				switch(in[5]){
+				case onechar('I', 0, 8):
+					switch(in[6]){
+					case onechar('R', 0, 8):
+						return enum_node_direction::DEC_DIR;
+					break;
+					default: break;
+					}
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		case onechar('I', 0, 32) | onechar('N', 8, 32) | onechar('C', 16, 32) | onechar('_', 24, 32):
+			switch(in[4]){
+			case onechar('D', 0, 8):
+				switch(in[5]){
+				case onechar('I', 0, 8):
+					switch(in[6]){
+					case onechar('R', 0, 8):
+						return enum_node_direction::INC_DIR;
+					break;
+					default: break;
+					}
+				break;
+				default: break;
+				}
+			break;
+			default: break;
+			}
+		break;
+		default: break;
+		}
+		break;
+	default: break;
+	}
+	if(throw_on_invalid)
+		throw std::runtime_error("Found unrecognized enum value " + std::string(in) + " of enum_node_direction.");
+	return enum_node_direction::UXSD_INVALID;
+}
+
+
+inline atok_t_channel alex_t_channel(const char *in){
```

Comment:
What is the difference between `alex` and `lex`?

Reply:
`lex` means lex an enum value, `alex` means lex an attribute name. I should make it more clear.

Reply:
`enum_lex` and `attribute_lex` (or maybe `lex_enum` and `lex_attribute`?)